### PR TITLE
Check broken links

### LINKED ISSUE #8107: Testing URL links in docs & readme
## 🚀 Feature

it may easily happen that some external links are not valid anymore so we shall test them all...

### Motivation

detect broken links in the process

### Alternatives

load the URLs via bs4? (a web scraping tool)
assert that the text you get back contains an expected string?

### PR DESCRIPTION
**Pull request by shenoynikhil**
## What does this PR do?

Fixes #8107
This checks for broken links in all markdown files.
- [x] For markdown files, I have created a new workflow based on [this](https://github.com/gaurav-nelson/github-action-markdown-link-check)
- [x] Skips checks on links of the form, `^https://github.com/Lightning-AI/lightning/pull/.*` (to ignore checks in CHANGELOG) 
- [x] There was an azure link (in `README.md` and `src/pytorch_lightning/README.`md) with `(` and `)` in the link which caused the `mdformat` pre-commit hook to update the link with `<` and `>`. This causes the md-link-check to fail. Update the link with replacing `(` with `$28` and `)` with `$29` encoding. This resolves it.
- [x] Added a json file in the `workflows/` directory to allow for modifying the parameters of the markdown link check. Currently situated at [markdown.links.config.json](https://github.com/Lightning-AI/lightning/pull/16168/files#diff-385b43dfa854b53c3a86b3c5ccd92246d21bc40ca2e81bfb178ca97a1048ae87), it contains some configuration to allow for loading of github links. 
- [x] Fixed two links that were not fixed in the #16191 
- [x] Fabric Doc Links were incorrect in multiple places, fixed them.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: .github/workflows/docs-checks.yml [hunk 1]
```diff
@@ -157,7 +157,8 @@ jobs:
       - name: Make Documentation
         working-directory: ./docs/source-${{ matrix.pkg-name }}
         run: |
-          make html --debug --jobs $(nproc) SPHINXOPTS="-W --keep-going"
+          make clean
```
Comment (Borda) [inline .github/workflows/docs-checks.yml:160]:
```suggestion
```
it shall alwyas start with clean

File: .github/workflows/md-check-links.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      - "!**/*.md"
+
+jobs:
+  markdown-link-check:
+    runs-on: ubuntu-20.04
+    steps:
+    - uses: actions/checkout@master
+    - uses: gaurav-nelson/github-action-markdown-link-check@v1
+      with:
+        use-quiet-mode: 'yes'
```
Comment (Borda) [inline .github/workflows/md-check-links.yml:28]:
```suggestion
        use-verbose-mode: 'yes'
```

File: .github/workflows/md-check-links.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    paths:
+      - "!*.md"
+      - ".github/**"
+      - "src/**"
+      - "tests/**"
+      - "requirements/**"
+      - "setup.py"
+      - "examples/**"
+      - "dockers/**"
+      - "!**/*.md"
```
Comment (Borda) [inline .github/workflows/md-check-links.yml:19]:
```suggestion
      - ".github/workflows/md-check-links.yml"
      - "**/*.md"
```
as it MD action we shall not skip MD files :)

Comment (shenoynikhil) [reply .github/workflows/md-check-links.yml]: Makes sense. I have added `**/*.md` but why do you suggest `- ".github/workflows/md-check-links.yml"`?

File: .github/workflows/docs-checks.yml (hunk 1 above)
Comment (shenoynikhil) [reply .github/workflows/docs-checks.yml]: Resolved this.

File: .github/workflows/md-check-links.yml (hunk 3 above)
Comment (shenoynikhil) [reply .github/workflows/md-check-links.yml]: I'm curious to know why `- ".github/workflows/md-check-links.yml"` is needed.

Comment (carmocca) [reply .github/workflows/md-check-links.yml]: Because you want changes to this file to trigger itself

**Review by Borda — APPROVED** [review]

**Comment by shenoynikhil** [comment]
I think there are still a few links that are showing some error (based on this [workflow job](https://github.com/Lightning-AI/lightning/actions/runs/3769618684/jobs/6408944285)). Will try to fix them in this PR. @Borda

**Comment by shenoynikhil** [comment]
@Borda It is ready for review. Have updated the PR description to update the changes I made after the #16191 was merged.

File: .github/workflows/docs-checks.yml [hunk 4]
```diff
@@ -156,7 +156,7 @@ jobs:
       - name: Make Documentation
         working-directory: ./docs/source-${{ matrix.pkg-name }}
         run: |
-          make html --debug --jobs $(nproc) SPHINXOPTS="-W --keep-going"
+          make html --debug --jobs $(nproc) SPHINXOPTS="-W --keep-going" -b linkcheck
```
Comment (awaelchli) [inline .github/workflows/docs-checks.yml:159]: How reliable is this? In my experience, link checkers can be flaky or produce false positives. Let's make sure this does not block code contributions.

**Review by awaelchli — COMMENTED** [review]
Let's make sure this is a non-required check so it does not block PRs from getting merged.

File: .github/workflows/docs-checks.yml (hunk 4 above)
Comment (shenoynikhil) [reply .github/workflows/docs-checks.yml:159]: I updated this based on how its done in [torchmetrics](https://github.com/Lightning-AI/metrics/blob/d2ef383c16f52f02ff38e7c7b8146010e694036b/.github/workflows/docs-check.yml#L79).

Comment (Borda) [reply .github/workflows/docs-checks.yml:159]: Let's move this docs build argument to the second PR where all links are fixed...

Comment (shenoynikhil) [reply .github/workflows/docs-checks.yml:159]: I'll remove it for now then?

Comment (awaelchli) [reply .github/workflows/docs-checks.yml:159]: If it can be done in a separate check that we mark optional, that would be great :)

Comment (shenoynikhil) [reply .github/workflows/docs-checks.yml:159]: Would it be like another workflow `docs-check-with-linkcheck` but that workflow is optional?

Comment (shenoynikhil) [reply .github/workflows/docs-checks.yml]: I have removed it from `docs-checks.yml` for now.

Comment (shenoynikhil) [reply .github/workflows/docs-checks.yml]: @awaelchli Should I add something like this in the same workflow, not sure if this is how we add an optional check
```
      - name: Make Documentation with Link Check (Optional)
        working-directory: ./docs/source-${{ matrix.pkg-name }}
        run: |
          make html --debug --jobs $(nproc) SPHINXOPTS="-W --keep-going" -b linkcheck
        continue-on-error: true
```

Comment (awaelchli) [reply .github/workflows/docs-checks.yml]: Can't really answer these questions. Better to ask @Borda. 
On a high-level, what I'm saying is that we probably want this check to just be informative, but not be a real failure that prevents from merging PRs.

Comment (Borda) [reply .github/workflows/docs-checks.yml]: yes, this check sometimes produce false positive, especially while checking GH PRs from the changelog, but I would still keep it - just check if we can exclude some files like the changelog from validation...
as for `continue-on-error: true` no as this job is mainly for validation that the docs can build so allowing to continue would not bring anything...

Comment (Borda) [reply .github/workflows/docs-checks.yml]: so let's keep this PR as it is and create another including adding -b link-check
 OR (better) set link check as a separate step where we allow continue on error
```
$ cd docs
$ make linkcheck
```

**Comment by Borda** [comment]
@shenoynikhil could you pls add this workflow also to https://github.com/Lightning-AI/utilities
cc: @akihironitta

**Review by akihironitta — APPROVED** [review]
Hey @shenoynikhil! Thanks for working on this :) LGTM!

Once we're confident that this action works properly in the future, we can include this in the probot's checkgroup config:
https://github.com/Lightning-AI/lightning/blob/859a228a915893f14759dda6d35f6050fe6df382/.github/checkgroup.yml

**Comment by akihironitta** [comment]
> @shenoynikhil could you pls add this workflow also to https://github.com/Lightning-AI/utilities cc: @akihironitta

@shenoynikhil Just created an issue here: https://github.com/Lightning-AI/utilities/issues/80 Feel free to comment on the issue if you're interested in working on it 💜

**Comment by carmocca** [comment]
> Once we're confident that this action works properly in the future, we can include this in the probot's checkgroup config

Since this check (`markdown-link-check`) is not really specific to any group, it would be simpler to mark it as "Required" directly.

**Review by carmocca — APPROVED** [review]
LGTM.

The description should be updated now that the sphinx changes were removed. I would also keep the linked issue open until those are checked too.

You might also consider adding the job to the utilities repository before merging this. Or you can merge it as-is and refactor it later to use the utilities repo copy

File: .github/workflows/docs-checks.yml (hunk 4 above)
Comment (shenoynikhil) [reply .github/workflows/docs-checks.yml]: Resolving this conversation.

**Comment by shenoynikhil** [comment]
I'm not sure why the App.cloud checks are failing?

**Comment by carmocca** [comment]
They are failing in master. We will rebase this PR once they failures are resolved.

**Comment by akihironitta** [comment]
Thanks for working on this @shenoynikhil! Since we already started https://github.com/Lightning-AI/utilities/pull/81, let's reuse that reusable workflow here as @carmocca suggests in https://github.com/Lightning-AI/lightning/pull/16168#pullrequestreview-1234818786:
> You might also consider adding the job to the utilities repository before merging this.

**Comment by Borda** [comment]
> let's reuse that reusable workflow here as @carmocca suggests in [#16168 (review)](https://github.com/Lightning-AI/lightning/pull/16168#pullrequestreview-1234818786)

it means we need to release it...

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**